### PR TITLE
add distinct operator

### PR DESCRIPTION
**Pull request by hillin**

### CODE REVIEW

File: API.md
Comment (conartist6) [inline API.md]: You'll want to make these examples smaller and simpler. Not everyone that reads these docs will even know english well, so it's pretty critical that it's easy to look at the example and understand it. I don't want people to have to do modulo 7 math in their head in order to understand how the method works.

Second, I think the example shows something incorrect about the implementation. I would expect:
```js
distinct((x) => x % 7, [14, 21, 9, 777]); // Iterable[14, 9]
```

File: src/impls/$distinct/$distinct.js
Comment (conartist6) [inline src/impls/$distinct/$distinct.js]: These numbers refer to the number of configuration arguments, i.e. those that are not iterables. You only have 1 arguments that is not an iterable, so `maxArgs: 1`

File: src/impls/$distinct/$distinct.js
Comment (conartist6) [inline src/impls/$distinct/$distinct.js:6]: The implementation looks good!

File: src/impls/$distinct/DOCME.json
Comment (conartist6) [inline src/impls/$distinct/DOCME.json]: You don't need to define `signatures` -- the documentation engine will infer it in all but the weirdest cases. I'm guessing it just did something weird because `maxArgs` was off.

File: src/impls/$distinct/__tests__/$distinct.test.ts
Comment (conartist6) [inline src/impls/$distinct/__tests__/$distinct.test.ts]: Let's make these objects one-liners so that more fits on the screen.

File: src/impls/$distinct/__tests__/$distinct.test.ts
Comment (conartist6) [inline src/impls/$distinct/__tests__/$distinct.test.ts]: This is OK as tests go, but I'd prefer that you use the smallest inputs that demonstrate particular behavior. I might do something like:
```
When input is empty: yields no values
null -> iter[]

When input contains only distinct values: yields values
iter[1] => iter[1]

When the input contains identical values: emits only the first value
iter[1, 1] => iter[1]
iter[1, 0, 1] => iter[1, 0]
```

It may look like this "doesn't test as much" as a more complex test case, but code coverage will tell you if you really hit all the cases.

**Review by conartist6 — CHANGES_REQUESTED** [review]
And lastly you'll need to update `CHANGELOG`.

Overall it looks quite good! Most of my comments are just little things.

File: API.md
Comment (conartist6) [reply API.md]: It does look like the implementation is correct, it's just the example output that's wrong.

Comment (hillin) [reply API.md]: > distinct((x) => x % 7, [14, 21, 9, 777]); // Iterable[14, 9]

Nah, all of 14, 21 and 777 has a modulo 7 result of 0. Modulo gives you the remainder after dividing.

I'll try to find simpler yet still meaningful examples.

File: src/impls/$distinct/$distinct.js
Comment (hillin) [reply src/impls/$distinct/$distinct.js]: Sure. The currying code is a little bit confusing but yet quite smart.

File: API.md
Comment (conartist6) [reply API.md]: Right. Meaning that of those numbers only 14 should appear in the output, because the other numbers are considered identical to it. 9 also makes it in because it's the first value for which the `x % 7 === 2`.

Comment (conartist6) [reply API.md]: The output can only contain values that were in the input (not 0 or 2).

File: src/impls/$distinct/$distinct.js
Comment (conartist6) [reply src/impls/$distinct/$distinct.js]: I should probably add some better comments to it.

File: API.md
Comment (hillin) [reply API.md]: Oh, you are definitely right.

Comment (conartist6) [reply API.md]: I actually think a simpler version of that modulo example might be an excellent one to use since it demonstrates that property which you've just proven may seem surprising.

Comment (hillin) [reply API.md]: fixed in 45b72934c52c0df4e907d05d07b1429a8bc7dc43

File: src/impls/$distinct/__tests__/$distinct.test.ts
Comment (hillin) [reply src/impls/$distinct/__tests__/$distinct.test.ts]: fixed in 45b72934c52c0df4e907d05d07b1429a8bc7dc43

File: src/impls/$distinct/__tests__/$distinct.test.ts
Comment (hillin) [reply src/impls/$distinct/__tests__/$distinct.test.ts]: fixed in 45b72934c52c0df4e907d05d07b1429a8bc7dc43

File: src/impls/$distinct/$distinct.js
Comment (hillin) [reply src/impls/$distinct/$distinct.js]: fixed in 64f50880549afd41939f95942a41006ffd65ce5d

File: src/impls/$distinct/DOCME.json
Comment (hillin) [reply src/impls/$distinct/DOCME.json]: fixed in 64f50880549afd41939f95942a41006ffd65ce5d

**Comment by hillin** [comment]
`CHANGELOG` updated in 64f50880549afd41939f95942a41006ffd65ce5d

File: API.md
Comment (conartist6) [inline API.md:447]: Sorry to ask for this one last change, but I'm pretty picky about my docs. I think the docs should say this:
```
Defaults:
- `selector`: `value => value`

Yields values from `source` with duplicate values eliminated. A value is considered a duplicate if `Object.is(selector(a), selector(b))`. A `Set` is used to do this check in constant time.
```

**Review by conartist6 — APPROVED** [review]
I'm approving this because it is usable, but per the comment there's one more change I want before I merge.

File: API.md
Comment (conartist6) [reply API.md:447]: What do you think is the best way to handle stuff like this? I guess I could just edit your branch if you've granted me perms. Unfortunately it seems like as a maintainer it's difficult to tell if you've been granted perms to edit PRs. It seems you can find out by opening the online text editor and if you don't have perms it'll tell you that you'd have to make changes in a fork. That's kind of dumb, but OK. It looks like you've given me those perms. Is it good etiquette to use them? Rather than just making changes you don't know about and merging them I'd rather be opening a PR to your PR, and then if you see my changes and agree with them we could merge both at the same time. I've never tried doing that though.

Comment (hillin) [reply API.md:447]: I'm totally okay if you push to my fork, I didn't do anything to grant you the permission so I guess it's Github's default policy.

> A `Set` is used to do this check in constant time

This is not accurate, the `has()` method of `Set` works in logarithmic time (`O(logn)`), not constant time (`O(1)`).

Comment (conartist6) [reply API.md:447]: What's your source for `logn` time?

Comment (hillin) [reply API.md:447]: Sorry, that's embarrassing, I have to take that one back. I do understand that typical (Hash)Set implementations have a time complexity of `O(1)`, but somehow I remember someone has made a benchmark and proved that it was actually `O(logn)`. But now I turn to look for that post I can't find it. Guess my memory betrayed me!

Comment (conartist6) [reply API.md:447]: I mean that's the thing with JS -- none of the perf is guaranteed. The spec doesn't say one way or the other. I don't actually have any evidence for `O(1)` either, but that's what I was taught about how hash maps should work. Anyway OK, let's update the docs for now and push this out so you can use it.

Comment (hillin) [reply API.md:447]: Yeah specs say it's sublinear but I guess it's not a good idea to use that word in the docs.

Are you gonna make the changes or should I?

Comment (conartist6) [reply API.md:447]: I'll do it tomorrow

Comment (hillin) [reply API.md:447]: ping

**Comment by conartist6** [comment]
Accepted as 5f6758

**Comment by conartist6** [comment]
I couldn't merge this PR itself because of the branch protection rules on your `trunk` -- I couldn't update that branch in your repo because it would rewrite history. That's of course something that your repo inherited from this repo, but that's generally exactly what I want so it was easier to just merge the commit manually. In the future I'll try to make sure that PRs are made from unprotected branches.

**Comment by conartist6** [comment]
Released in `v7.4.0`

**Pull request closed without merging**
